### PR TITLE
2015 day06 branch

### PR DESCRIPTION
**Pull request by TheAndi1**
first part of day06 challenge

### CODE REVIEW

File: day06_solution.py [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+
+
+
+content = Path('day06_input.txt').read_text()
+process_text_input(content)
+
+
+assert(process_text_input("turn on 0,0 through 999,999")) == 1000000
+assert(process_text_input("toggle 0,0 through 999,0")) == 1000
+assert(process_text_input("turn off 499,499 through 500,500")) == 0
```
Comment (dwt) [inline day06_solution.py]:
```suggestion
assert process_text_input("turn on 0,0 through 999,999") == 1000000
assert process_text_input("toggle 0,0 through 999,0") == 1000
assert process_text_input("turn off 499,499 through 500,500") == 0
```

assert ist keine Funktion, sondern ein keyword, daher schreibt man das normalerweise so. Warum ist das ein Keyword: Weil es ein Feature in python gibt, `-O` die alle Assert statements überspringt - das würde bei einer Funktion auch funktionieren, aber nicht so viel zeit sparen, weil der Funktionsaufruf - in deinem Beispiel `process_text_input("turn off 499,499 through 500,500")` trotzdem berechnen würde und ja schon sehr teuer sein kann.

File: day06_solution.py [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+    return -1
+
+
+def split_text_line(text_line):
+    process = text_line[:find_first_number(text_line) - 1]
+    text_line = text_line[find_first_number(text_line):]
+    coord1 = (int(text_line[:find_first_comma(text_line)]), int(text_line[find_first_comma(text_line) + 1:find_first_space(text_line)]))
+    text_line = text_line[find_first_space(text_line)+1:][find_first_space(text_line[find_first_space(text_line) + 1:]) + 1:]
+    coord2 = (int(text_line[:find_first_comma(text_line)]), int(text_line[find_first_comma(text_line) + 1:]))
+    return process, coord1, coord2
```
Comment (dwt) [inline day06_solution.py]: Das funktioniert ganz gut, ich würde das aber eher so Parsen, weil ich das sehr viel kürzer finde:

```python
import re

def parse_instructions(instruction):
    pattern = r"(?P<action>turn on|turn off|toggle) (?P<x1>\d+),(?P<y1>\d+) through (?P<x2>\d+),(?P<y2>\d+)"
    match = re.match(pattern, instruction)
    
    if match:
        result = match.groupdict()
        # Convert coordinates to integers
        for key in ['x1', 'y1', 'x2', 'y2']:
            result[key] = int(result[key])
        return result
    return None

# Test cases
instructions = [
    "turn on 0,0 through 999,999",      {'action': 'turn on',   'x1': 0,    'y1': 0,    'x2': 999, 'y2': 999},
    "toggle 0,0 through 999,0",         {'action': 'toggle',    'x1': 0,    'y1': 0,    'x2': 999, 'y2': 0},
    "turn off 499,499 through 500,500", {'action': 'turn off',  'x1': 499,  'y1': 499,  'x2': 500, 'y2': 500},
]

# Test the parser
each = iter(instructions)
for instruction in iter(each):
    solution = next(each)
    parsed = parse_instructions(instruction)
    assert solution == parsed, f"Instruction: {instruction} Parsed: {parsed}"
```

Das erfordert natürlich reguläre ausdrücke, die zu verstehen ist aber ein _sehr_ gutes investment, weil die Ständig vorkommen und überall (auch außerhalb der Programmierung) sehr anwendbar sind.

File: day06_solution.py [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+                            light_grid[row_index][light_index] = 1
+                        case "turn off":
+                            light_grid[row_index][light_index] = 0
+                        case "toggle":
+                            if light == 0:
+                                light_grid[row_index][light_index] = 1
+                            elif light == 1:
+                                light_grid[row_index][light_index] = 0
+                        case _:
+                            print("invalid")
```
Comment (dwt) [inline day06_solution.py]: Das funktioniert so auf jeden fall und verwendet die Python tools die Du schon kennst. Kudos dafür!

Wenn Du allerdings numpy verwendest, dann macht es auch Sinn die Funktionen davon zu verwenden. Damit kannst Du das Koordinatensystem nämlich auch so ausdrücken und dir ne menge Tipp- und Rechen-Arbeit sparen:

```python
import numpy as np

def display_bool_matrix(heading, matrix, true_char='█', false_char='·'):
    print(heading)
    for row in matrix:
        print(''.join(true_char if cell else false_char for cell in row))
    print('lamps on', np.sum(matrix))


matrix = np.zeros((10, 10), dtype=bool)

x1, y1 = 2, 3
x2, y2 = 4, 5

# set rectangle to True, but could just as well be False for 'turn off'
matrix[x1:x2+1, y1:y2+1] = True  # +1 because coordinates are inclusive

display_bool_matrix("middle on", matrix)

# Toggle all values in a sub-rectangle: Using XOR with True
matrix[x1:x2+1, y1:y2+1] ^= True  # +1 because coordinates are inclusive

display_bool_matrix("now toggled off again", matrix)
```

**Review by dwt — COMMENTED** [review]
Nice!

Ich vermute das Du beim Parsen und anwenden relativ lange gebraucht hast, weil Du für Tests unten nur die drei 'ende zu ende' tests hattest. Wen dem nicht so war, krass Du bist schon jetzt ein besserer Programierer als ich. :)

Falls doch, ich würde empfehlen für die einzelnen teile: Parsen, anwenden der operationen Ausgabe, separat zu testen, damit Du die verschiedenen edge cases leichter zusammen kriegst.

Und natürlich würde ich dir gerne pytest und die test-funktionen nahe legen weil dir das noch mal besseres feedback gibt wenn etwas schief geht.

File: day06.2_solution.py [hunk 4]
```diff
@@ -0,0 +1,67 @@
+from __future__ import annotations
+
+import re
+from typing import Iterable, Tuple, List
+
+# --------------------------------------------------------------------------- #
+# 1. RegEx – Einmal kompilieren (ein Pattern reicht)
+# --------------------------------------------------------------------------- #
+_CMD_RE = re.compile(
+    r"^(turn )?(on|off|toggle) (\d+),(\d+) through (\d+),(\d+)$"
```
Comment (dwt) [inline day06.2_solution.py:10]: Du hast keine benannten Gruppen verwendet, magst Du kurz erklären wieso dir das so besser gefallen hat?

Einige Details zu Regexes die Du vermutlich noch nicht kennst:

[`(?:....)` ist ein 'non capturing group' ](https://docs.python.org/3.13/library/re.html#regular-expression-syntax) - d.h. sie taucht nicht in `.group()` oder `.groupdict()` auf und stört daher weniger. Wenn du etwas also nur in eine Gruppe verpackst um es optional zu machen (mit dem ? danach) dann ist das ein sehr gute option das gleich so auszuzeichnen. (Dadurch ändern sich natürlich deine Offsets, was der Grund für die Benannten Gruppen ist, da man damit keine Offsets mehr hat. :)

File: day06.2_solution.py [hunk 5]
```diff
@@ -0,0 +1,67 @@
+from __future__ import annotations
+
+import re
+from typing import Iterable, Tuple, List
+
+# --------------------------------------------------------------------------- #
+# 1. RegEx – Einmal kompilieren (ein Pattern reicht)
+# --------------------------------------------------------------------------- #
+_CMD_RE = re.compile(
```
Comment (dwt) [inline day06.2_solution.py:9]: Ich mag, dass du den `re.compile()` rausgezogen hast, das ist echt gute Praxis.

File: day06.2_solution.py [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+                if action == "on":
+                    row[y] += 1
+                elif action == "off":
+                    if row[y] > 0:
+                        row[y] -= 1
+                else:  # toggle
+                    row[y] += 2
+
+    # Summe aller Werte im Raster
+    return sum(sum(row) for row in lights)
```
Comment (dwt) [inline day06.2_solution.py:43]: wenn du weiter auf numpy setzen würdest, dann würde das hier noch mal krass zusammenschrumpfen, weil du die range Operatoren mit universal functions super kombinieren könntest. aber das ist wunderschön einfach - was auch ein Vorteil ist.

File: day06.2_solution.py [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+    assert total_brightness(
+        [
+            "turn on 0,0 through 0,0",
+            "toggle 0,0 through 0,0",
+            "turn off 0,0 through 0,0",
+        ]
+    ) == 2
+    assert total_brightness(["turn off 0,0 through 0,0"]) == 0
+    print("Alle Tests erfolgreich!")
+
```
Comment (dwt) [inline day06.2_solution.py:58]: sehr wenig code, und auch sehr wenig tests. Wie war denn deine Erfahrung beim Schreiben? Haben die tests gereicht um jederzeit ein gutes flow Erlebnis beim coden zu haben, oder hättest Du den code noch mehr in Funktionen aufteilen sollen um diese einzeln testen zu können?

Beim Zweiten mal geht es ja meistens einfacher, weil man das problem besser verstanden hat, daher gibt es hier vermutlich keine klare Antwort.

File: day06.2_solution.py [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+                    if row[y] > 0:
+                        row[y] -= 1
+                else:  # toggle
+                    row[y] += 2
+
+    # Summe aller Werte im Raster
+    return sum(sum(row) for row in lights)
+
+
+def _run_tests() -> None:
```
Comment (dwt) [inline day06.2_solution.py:46]: Besser wäre es den test `test_total_brightness()` zu nennen. (Warum kommt gleich unten)

File: day06.2_solution.py [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+            "toggle 0,0 through 0,0",
+            "turn off 0,0 through 0,0",
+        ]
+    ) == 2
+    assert total_brightness(["turn off 0,0 through 0,0"]) == 0
+    print("Alle Tests erfolgreich!")
+
+
+if __name__ == "__main__":
+    _run_tests()  # Selbst‑tests
```
Comment (dwt) [inline day06.2_solution.py:61]: Ich würde die tests mit `watching_testrunner pytest day06.2_solution.py` ausführen. Wenn Du die tests `test_$something` benennst, werden die damit automatisch ausgeführt, und Du musst die nicht in der main Methode mit ausführen, das macht die Entwicklung leichter und schneller.

File: day06.2_solution.py [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+    try:
+        with open("day06_input.txt") as f:
+            data = f.read().strip().splitlines()
```
Comment (dwt) [inline day06.2_solution.py:64]: Consider: 

```python
from pathlib import Path
data = Path('day06.2_solution.py').read_text().strip().splitlines()
```

Das ist der modernere weg mit (kleinen) Dateien zu Arbeiten.
